compiler/rustc_span/src/source_map.rs

```diff
@@ -127,30 +127,16 @@ pub struct StableSourceFileId(u128);
 // StableSourceFileId, perhaps built atop source_file.name_hash.
 impl StableSourceFileId {
     pub fn new(source_file: &SourceFile) -> StableSourceFileId {
-        StableSourceFileId::new_from_pieces(
-            &source_file.name,
-            source_file.name_was_remapped,
-            source_file.unmapped_path.as_ref(),
-        )
+        StableSourceFileId::new_from_name(&source_file.name)
     }
 
-    fn new_from_pieces(
-        name: &FileName,
-        name_was_remapped: bool,
-        unmapped_path: Option<&FileName>,
-    ) -> StableSourceFileId {
+    fn new_from_name(name: &FileName) -> StableSourceFileId {
         let mut hasher = StableHasher::new();
 
-        if let FileName::Real(real_name) = name {
-            // rust-lang/rust#70924: Use the stable (virtualized) name when
-            // available. (We do not want artifacts from transient file system
-            // paths for libstd to leak into our build artifacts.)
-            real_name.stable_name().hash(&mut hasher)
-        } else {
-            name.hash(&mut hasher);
-        }
-        name_was_remapped.hash(&mut hasher);
-        unmapped_path.hash(&mut hasher);
+        // If name was remapped, we need to take both the local path
+        // and stablised path into account, in case two different paths were
+        // mapped to the same
+        name.hash(&mut hasher);
```

Comment:
I'll need to look into this some more. We don't want to revive https://github.com/rust-lang/rust/issues/70924 again.

Reply:
OK, it looks to me like the previous version of this code only fixed #70924 because `try_to_translate_virtual_to_real()` in `CrateMetadataRef::imported_source_files()` does not correctly update `unmapped_path` and `name_was_remapped`. In order to preserve the fix we must hash only the remapped path. Maybe we could (manually) implement `Hash` for the proposed `FileNameDisplay` type for this purpose.

Reply:
Good catch, this probably would've reintroduced #70924. I manually implemented `Hash` for `RealFileName` so it hashes only the remapped path if available.

Still concerned about the edge case of two files mapping to the same thing. E.g. we have a `src/lib.rs` relative to the current working directory, then there's a `/dependency/src/lib.rs` and we do `--remap-path-prefix=/dependency/=`, causing the latter to be remapped to the former. 

I still don't understand why change in presence of `rust-src` would trigger a panic. If `rust-src` was present before and the compilation cached, then we remove the component, surely it would lead to a different `StableSourceFileId` if we just hash the whole `RealFileName` with `local_path` (now is `None`), causing the cache to be rebuilt?